Add Duplicate Lines shortcut to CodeTextEditor

PucklaJ opened the pull request:
This PR adds the shortcut `script_text_editor/duplicate_lines`. The default buttons are `Ctrl+Alt+Down`. For macOS the default buttons are `Cmd+Shift+Down`. These shortcuts are similiar to the VS Code command `editor.action.copyLinesDownAction`. If no text is selected it behaves the same as `duplicate_selection`, but if text is selected it will duplicate the whole lines of the selection.
This also supports multiple carets.

`script_text_editor/duplicate_lines`:
![duplicate_lines_below](https://user-images.githubusercontent.com/25499721/194751129-b8001410-5968-4097-b788-3f8dbcfa0b23.gif)

`script_text_editor/duplicate_selection`:
![duplicate_selection](https://user-images.githubusercontent.com/25499721/194751138-7c499b89-550b-4366-8195-e565234a52a8.gif)

Commit:
Add Duplicate Lines shortcut to CodeTextEditor

This keyboard shortcut has been made with inspiration from the VS Code keyboard shortcut editor.action.copyLinesDownAction. It duplicates all selected lines and inserts them below no matter where the caret is within the line.

PucklaMotzer09 commented:
Added suport for multiple carets

MewPurPur commented:
I'll do a code review in a sec, but can you argue for why you think this feature should be added? We don't accept features just because PRs exist for them.

I personally think Ctrl+Shift+D does what I want just fine. If I have no selections, this PR does the same, just adjusts the caret in one of the scenarios. If I want to duplicate multiple lines, I'm just going to do a proper selection with triple-clicking and then dragging, both very well-known functionalities.

MewPurPur commented on editor/code_editor.cpp:
```diff
@@ -1608,6 +1618,94 @@ void CodeTextEditor::duplicate_selection() {
+		text_editor->set_caret_line(new_caret_line, true, true, 0, caret_index);
+		text_editor->set_caret_column(new_caret_column, true, caret_index);
+
+		if (select_from_line != -1) {
+			if (p_below) {
+				// Advance the selection by the number of duplicated lines
+				select_from_line += select_num_lines;
+				select_to_line += select_num_lines;
+			}
+			text_editor->select(select_from_line, select_from_column, select_to_line, select_to_column, caret_index);
+		}
+	}
```
Likewise, just put this stuff in the `has_selection()` if/else.

PucklaJ replied:
I unfortunately can not put this before inserting the text, because I need to deselect when inserting text, because the inserted text replaces the selection.

PucklaJ replied:
If I could use `_insert_text`, which is private, I could put this before the text insertion, because it wouldn't affect the selection.

MewPurPur commented on editor/code_editor.cpp:
```diff
@@ -1608,6 +1618,94 @@ void CodeTextEditor::duplicate_selection() {
+		// The new column position of the caret after the operation
+		int new_caret_column = text_editor->get_caret_column(caret_index);
+		// The caret positions of the selection. Stays -1 if there is no selection
+		int select_from_line = -1;
+		int select_to_line = -1;
```
You don't use these much outside of places where `has_selection()` is checked. You should define them close to the scope where the variables have a meaning.

PucklaJ replied:
If I put them in their own scope I would need to write duplicate code.

MewPurPur reviewed (changes requested):
Haven't tested the feature itself, just a code review.

torcado194 commented:
I would love this feature. I have never once wanted to duplicate a subsection within a line using a hotkey, *especially* if that selection spans multiple lines, I couldn't imagine a scenario in which I would want to duplicate that content in-place.
I duplicate lines constantly, duplicating and swapping lines are probably my most used hotkeys aside from undo/redo. Needing to make sure my selections encompass the entirety of the lines I want to duplicate before i'm able to duplicate multiple lines is very tedious, especially with no way to automatically spread selections to the start/end of lines. If a cursor in my multi-cursor selection started in the middle of the line? Now i have to redo the whole selection to move the start point to the beginning of the line.
And a mouse-based alternative isn't ideal. I'm already typing, I want to use the keyboard, not the mouse. At least for something this simple.

MewPurPur commented:
@torcado194 That makes sense, I think "Duplicate Lines" is a hard thing to argue against. What about this particular implementation, with separate "below" and "above" versions? That's the main part I find overly complicated.

torcado194 commented:
I agree, I hardly ever use "duplicate above". "below" is the default direction for the duplicate line action for any editor I use (and is what I'd expect). the "duplicate above" result can be easily substituted using the keyboard by duplicating below and swapping the new selected lines upward, at the cost of additional keystrokes equal to the line count.
but on the other hand the duplicate above and below functions are found in many editors I've used (visual studio and vs code for example)
maybe it's not part of godot's philosophy but I would *much* rather have more hotkey functions than less, even if they're unassigned by default.

PucklaJ commented:
> What about this particular implementation, with separate "below" and "above" versions?

I tried to recreate the feature from VS Code and there they have both a above and below version. But I agree I hardly use it either.

PucklaJ commented on editor/code_editor.cpp:
```diff
@@ -1608,6 +1618,94 @@ void CodeTextEditor::duplicate_selection() {
+		// The caret positions of the selection. Stays -1 if there is no selection
+		int select_from_line = -1;
+		int select_to_line = -1;
+		int select_from_column = -1;
+		int select_to_column = -1;
```
I tried to remove them, but when the caret moves, the selection gets unset, which means that I need to store them before changing the caret. And I think that I should transfer the selection to the duplication since it means that you can perform the same duplication multiple times in a row. Which is something I use very often.

PucklaJ commented:
I will merge all commits into one, once you are done reviewing

PucklaJ commented:
@MewPurPur @Paulb23 This PR has just been hanging here for quite some time. Can you please finish reviewing it. I would really like it to be merged. I think it adds some good keyboard shortcuts which can really improve your productivity.

MewPurPur commented:
I still think it should be just "Duplicate Lines" that does it below, instead of above and below versions. It's not like regular duplicate is "before" and "after" I don't got much else to say, the code to me looks good and my remarks have been clarified or addressed.

PucklaJ commented:
Since it has been a few months since I last visited this PR I performed a rebase. On top of that I removed the duplicate_lines_above shortcut and only the below one stays with the short name duplicate_lines. Since duplicate_lines behaves the same as duplicate_selection without a selection I replaced it with duplicate_lines. In case you don't like this change I can just remove the last commit.
But I think that it makes sense to replace duplicate_selection, because I think that probably nobody has been using duplicate_selection with a selection. By replacing the old shortcut people would use the new one and notice the improvement.

MewPurPur commented:
All in all, I agree, I think this does great with preventing bloat. I've misused duplicate selection a few times and never found usecases for it. Other opinions would be great to gather, as I don't use IDEs and Godot's script editor is the only thing I have lots of experience with.

PucklaJ commented:
@Paulb23 What do you think?

MewPurPur commented on editor/code_editor.cpp:
```diff
@@ -1418,48 +1418,68 @@ void CodeTextEditor::delete_lines() {
+			for (int i = select_from_line; i <= select_to_line; i++) {
+				insert_text += "\n" + text_editor->get_line(i);
+			}
+			new_caret_line = select_to_line;
+
+			new_caret_line += select_num_lines;
```
`new_caret_line = select_to_line + select_num_lines;`

PucklaJ replied:
You are right. This has been overlooked since removing the duplicate_lines_above code.

MewPurPur reviewed (approved):
Looked at it again - no viruses have been added, the code is better, I couldn't think of anything missing from it. As already mentioned, I think only the change itself could possibly rustle some jimmies.

PucklaJ commented:
@Paulb23 @YuriSizov Could someone please also have a look at this PR and maybe even approve it? The PR has already deemed fine by @MewPurPur and the code itself is also okay. Could we please move on with this PR. It is just a key binding that would probably be a great addition for some users. Those you don't need or don't care about keybindings can just ignore it without any implications. I think there really is nothing to discuss anymore besides it replacing the `duplicate_selection` shortcut. Which I can just easily undo if this is an issue. Therefore I ask the Godot team to approve this PR and merge it.

YuriSizov commented:
Paul is the area maintainer, so I would still prefer his approval on this rather than doing it myself. There is also no proposal for this, so it's unclear to me if this is something that would benefit many users, or just a few. 

That said, before this can be merged you need to squash your commits.

PucklaJ commented:
> That said, before this can be merged you need to squash your commits.

Done

Paulb23 reviewed (commented):
Code looks fine, few minor comments.

I'm wanting to move things like this into `CodeEdit`, could then take advantage of unit tests as well. If you fancy doing that :)

PucklaJ commented:
@Paulb23 I moved the code into `CodeEdit` and then adjusted the code based on your review.

Paulb23 commented on scene/gui/code_edit.h:
```diff
@@ -465,6 +465,8 @@ class CodeEdit : public TextEdit {
 
 	void set_symbol_lookup_word_as_valid(bool p_valid);
 
+	void duplicate_lines();
```
`nit:` Can title this with a new section i.e from `TextEdit`:  `// Text manipulation` or `// Input actions`

Paulb23 reviewed (commented):
Thanks!

One small change, could you bind it to to the `CodeEdit` API? and again would appreciate any unit tests.

PucklaJ commented:
@Paulb23 After a rebase and fixing a merge conflict I added unit tests

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -2412,6 +2412,68 @@ Ref<Texture2D> CodeEdit::_get_folded_eol_icon() const {
 	return theme_cache.folded_eol_icon;
 }
 
+/* Text manipulation */
+void CodeEdit::duplicate_lines() {
```
Seems to be from #81551, but could you move this above `/* Visual */` so it's the same order as the header?

Otherwise looks great, really appreciate the unit tests too!

PucklaJ replied:
Done

AThousandShips commented on doc/classes/CodeEdit.xml:
```diff
@@ -152,6 +152,12 @@
+		<method name="duplicate_lines">
+			<return type="void" />
+			<description>
+				Duplicates all lines currently selected with any caret. Duplicates the entire line beneath the current one no matter where the caret is inside the line.
```
```suggestion
				Duplicates all lines currently selected with any caret. Duplicates the entire line beneath the current one no matter where the caret is within the line.
```
I believe more appropriate word

YuriSizov commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2395,6 +2399,8 @@ void ScriptTextEditor::register_editor() {
 	ED_SHORTCUT("script_text_editor/duplicate_selection", TTR("Duplicate Selection"), KeyModifierMask::SHIFT | KeyModifierMask::CTRL | Key::D);
 	ED_SHORTCUT_OVERRIDE("script_text_editor/duplicate_selection", "macos", KeyModifierMask::SHIFT | KeyModifierMask::META | Key::C);
+	ED_SHORTCUT("script_text_editor/duplicate_lines", TTR("Duplicate Lines"), KeyModifierMask::CMD_OR_CTRL | KeyModifierMask::ALT | Key::DOWN);
+	ED_SHORTCUT_OVERRIDE("script_text_editor/duplicate_lines", "macos", KeyModifierMask::CMD_OR_CTRL | KeyModifierMask::ALT | Key::Z);
```
Is this a common macOS shortcut for this purpose?

PucklaJ replied:
I don't know. Honestly I just put anything there because it would have been a duplicate otherwise

akien-mga replied:
Would be good to know what's the expected macOS shortcut for this operation (e.g. in VS Code or Xcode). CC @bruvzg @coppolaemilio

PucklaJ replied:
I just [looked it up](https://code.visualstudio.com/shortcuts/keyboard-shortcuts-macos.pdf). On VS Code its `shift+cmd+down`

akien-mga replied:
That might work. Seems like we use this mapping but only for PC here, and there's a different override for macOS:
```cpp
// core/input/input_map.cpp
        inputs = List<Ref<InputEvent>>();
        inputs.push_back(InputEventKey::create_reference(Key::DOWN | KeyModifierMask::SHIFT | KeyModifierMask::CMD_OR_CTRL));
        default_builtin_cache.insert("ui_text_caret_add_below", inputs);

        inputs = List<Ref<InputEvent>>();
        inputs.push_back(InputEventKey::create_reference(Key::L | KeyModifierMask::SHIFT | KeyModifierMask::CMD_OR_CTRL));
        default_builtin_cache.insert("ui_text_caret_add_below.macos", inputs);
```

PucklaJ replied:
I changed it on macos to `cmd_or_ctrl+shift+down`

YuriSizov reviewed (approved):
Consensus seems to be for this, at a glance the style looks fine.

YuriSizov commented:
Could you please squash your commits into one? Make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

PucklaJ commented:
@YuriSizov @Paulb23 @MewPurPur The commits are squashed. Everything is ready. Thank you for approving this PR.

akien-mga commented:
Thanks!
